aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,712 @@
+type PermitRx = oneshot::Receiver<Result<OwnedSemaphorePermit, Error>>;
+
+/// Handler-side ticket which is used to get in the queue.
+#[derive(Debug)]
+struct Ticket {
+    id: TicketId,
+    permit_rx: Option<PermitRx>,
+    cancel_tx: mpsc::UnboundedSender<DispatcherMessage>,
+}
+
+impl Ticket {
+    /// Waits for the ticket to be called. Returns an [`OwnedSemaphorePermit`] which should be
+    /// dropped to signal that the request is done.
+    ///
+    /// # Panics
+    ///
+    /// Panics if it is called more than once.
+    async fn wait(&mut self) -> Result<OwnedSemaphorePermit, Error> {
+        self.permit_rx
+            .take()
+            .unwrap()
+            .await
+            .map_err(|err| Error::Internal(format!("dispatcher task died: {}", err)))?
+    }
+}
+
+impl Drop for Ticket {
+    fn drop(&mut self) {
+        if self.permit_rx.is_some() {
+            let _ = self
+                .cancel_tx
+                .send(DispatcherMessage::Dequeue(self.id))
+                .map_err(|err| warn!("failed to send cancellation message: {:?}", err));
+        }
+    }
+}
+
+/// A handler that queues requests in a LIFO manner, according to the parameters set in a
+/// [`LIFORequestQueue`].
+///
+/// Multiple request handlers can share a queue, by cloning the [`Arc`] that wraps the queue.
+#[derive(Handler)]
+pub struct LIFOQueueHandler<H> {
+    #[handler(except = [run])]
+    handler: H,
+    queue: Arc<LIFORequestQueue>,
+}
+
+impl<H: Handler> LIFOQueueHandler<H> {
+    pub fn new(queue: Arc<LIFORequestQueue>, handler: H) -> Self {
+        Self { handler, queue }
+    }
+
+    async fn run(&self, mut conn: Conn) -> Conn {
+        match self.queue.acquire() {
+            Ok(mut ticket) => match conn.cancel_on_disconnect(ticket.wait()).await {
+                Some(permit) => match permit {
+                    Ok(_permit) => self.handler.run(conn).await,
+                    Err(err) => err.run(conn).await,
+                },
+                None => Error::ClientDisconnected.run(conn).await,
+            },
+            Err(err) => err.run(conn).await,
+        }
+    }
+}
+
+/// Convenience function for wrapping a handler with a [`LIFOQueueHandler`].
+pub fn queued_lifo<H: Handler>(queue: Arc<LIFORequestQueue>, handler: H) -> impl Handler {
+    LIFOQueueHandler::new(queue, handler)
+}
+
+struct Metrics {
+    /// The approximate number of requests currently being serviced by the queue. It's approximate
+    /// since the queue length may have changed before the measurement is taken. In practice, the
+    /// error should only be +/- 1. It is also more or less suitable for synchronization during
+    /// tests.
+    outstanding_requests: Gauge<u64>,
+}
+
+impl Metrics {
+    const OUTSTANDING_REQUESTS_METRIC_NAME: &'static str = "janus_aggregator_outstanding_requests";
+
+    fn new(meter: &Meter) -> Self {
+        // Note: this instrument's name is constant for the entire process. If we end up having a
```

Comment:
IMO, let's parameterize this now. IIRC it's slightly painful to introduce a parameter later.